server/plugins/LFUGens.cpp

```diff
@@ -2379,33 +2379,126 @@ void EnvGen_Ctor(EnvGen* unit) {
+    } break;
+    case shape_Linear: {
+        unit->m_grow = (endLevel - level) / counter;
+        // Print("grow %g\n", unit->m_grow);
+    } break;
+    case shape_Exponential: {
+        unit->m_grow = pow(endLevel / level, 1.0 / counter);
+    } break;
+    case shape_Sine: {
+        double w = pi / counter;
+
+        unit->m_a2 = (endLevel + level) * 0.5;
+        unit->m_b1 = 2. * cos(w);
+        unit->m_y1 = (endLevel - level) * 0.5;
+        unit->m_y2 = unit->m_y1 * sin(pi * 0.5 - w);
+        level = unit->m_a2 - unit->m_y1;
+    } break;
+    case shape_Welch: {
+        double w = (pi * 0.5) / counter;
+
+        unit->m_b1 = 2. * cos(w);
+
+        if (endLevel >= level) {
+            unit->m_a2 = level;
+            unit->m_y1 = 0.;
+            unit->m_y2 = -sin(w) * (endLevel - level);
+        } else {
+            unit->m_a2 = endLevel;
+            unit->m_y1 = level - endLevel;
+            unit->m_y2 = cos(w) * (level - endLevel);
+        }
+        level = unit->m_a2 + unit->m_y1;
+    } break;
+    case shape_Curve: {
+        if (fabs(curve) < 0.001) {
+            unit->m_shape = 1; // shape_Linear
+            unit->m_grow = (endLevel - level) / counter;
+        } else {
+            double a1 = (endLevel - level) / (1.0 - exp(curve));
+            unit->m_a2 = level + a1;
+            unit->m_b1 = a1;
+            unit->m_grow = exp(curve / counter);
+        }
+    } break;
+    case shape_Squared: {
+        unit->m_y1 = sqrt(level);
+        unit->m_y2 = sqrt(endLevel);
+        unit->m_grow = (unit->m_y2 - unit->m_y1) / counter;
+    } break;
+    case shape_Cubed: {
+        unit->m_y1 = pow(level, 1.0 / 3.0); // 0.33333333);
+        unit->m_y2 = pow(endLevel, 1.0 / 3.0);
+        unit->m_grow = (unit->m_y2 - unit->m_y1) / counter;
+    } break;
+    };
+    return true;
+}
+
 static bool check_gate(EnvGen* unit, float prevGate, float gate, int& counter, double level, int counterOffset = 0) {
     if (prevGate <= 0.f && gate > 0.f) {
         unit->m_stage = -1;
         unit->m_released = false;
         unit->mDone = false;
         counter = counterOffset;
         return false;
-    } else if (gate <= -1.f && prevGate > -1.f && !unit->m_released) {
-        // cutoff
-        int numstages = (int)ZIN0(kEnvGen_numStages);
+    } else if (gate <= -1.f && prevGate > -1.f) {
+        // forced release: jump to last segment overriding its duration
         float dur = -gate - 1.f;
         counter = (int32)(dur * SAMPLERATE);
         counter = sc_max(1, counter) + counterOffset;
-        unit->m_stage = numstages;
-        unit->m_shape = shape_Linear;
-        // first ZIN0 gets the last envelope node's level, then apply levelScale and levelBias
-        unit->m_endLevel = ZIN0(unit->mNumInputs - 4) * ZIN0(kEnvGen_levelScale) + ZIN0(kEnvGen_levelBias);
-        unit->m_grow = (unit->m_endLevel - level) / counter;
+        unit->m_stage = (int)ZIN0(kEnvGen_numStages) - 1;
```

Comment:
in general prefer the most restrictive cast  -- that would be `static_cast` here

Reply:
Ok, maybe you saw that, but I just copied the previous implementation, assigning directly to m_stage instead of a new variable. Anyway, I can change that cast! How about the one right above?

Reply:
oh i see! i was trying to be careful to style-review only code that you wrote, because you shouldn't be responsible for the code you just move around. i missed that you had copied the expression from a few lines above.

if you have the time and interest, please fix this one, but i'd prefer if you didn't do the one above, simply because that line wasn't modified in this diff and so modifying it now creates extra clutter. my practice is: if a line is modified in a commit, freely update its syntax/formatting for unambiguously better idioms, otherwise save those kinds of changes for separate commits, preferably making several identical changes in many places at once. in my experience, that makes it a lot easier to resolve merge conflicts and to browse git history.

along similar lines, ideally this PR should consist of a single squashed/amended commit, i.e. no extra commits addressing review comments. we don't go to the trouble of enforcing this or recommending it zealously, but that's my thoughts since we're on the topic. :)

Reply:
Sure, I thought so too, just wanted to make sure :) Thanks!
About squashing: shall I squash also the tests so that we end up with only one commit?

Reply:
no problem. i personally like to see code change + corresponding test change in a single commit, yes :)

Reply:
done!